### PR TITLE
Add modify name proposal

### PR DESCRIPTION
**Pull request by Taztingo**
## Description

Add a proposal to modify the name and restricted flag of a name.

closes: #1004

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote unit and integration [tests](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: proto/provenance/name/v1/name.proto [hunk 1]
```diff
@@ -47,6 +47,20 @@ message CreateRootNameProposal {
...
+message ModifyNameProposal {
+  option (gogoproto.equal)            = false;
+  option (gogoproto.goproto_getters)  = false;
+  option (gogoproto.goproto_stringer) = false;
+
+  string title       = 1;
+  string description = 2;
+  string name        = 3;
+  string owner       = 4;
+  bool   restricted  = 5;
```
Comment (SpicyLemon) [inline proto/provenance/name/v1/name.proto:61]: Could use some field comments here to indicate what's the supposed to be old/existing, and which fields are new to be applied.

File: x/name/client/cli/gov.go [hunk 2]
```diff
@@ -106,7 +110,93 @@ $ %s tx gov submit-proposal param-change tx gov submit-proposal \
 		},
 	}
 
-	cmd.Flags().String(FlagOwner, "", "The owner of the new name, optional (defaults to from address")
+	cmd.Flags().String(FlagOwner, "", "The owner of the new name, optional (defaults to from address)")
+	cmd.Flags().BoolP(flagRestricted, "r", true, "Restrict creation of child names to owner only, optional (default false)")
```
Comment (SpicyLemon) [inline x/name/client/cli/gov.go:114]: This `flagRestricted` needs some love both here and in the new `GetModifyNameProposalCmd ` function.

If you run `provenanced tx gov submit-legacy-proposal root-name-proposal --help`, you'll see this in the flags listing:
```
  -r, --restrict                 Restrict creation of child names to owner only, optional (default false) (default true)
```

1. The "(default false)" in the description is just flat out incorrect. And boolean flags should always be optional, so the whole `", optional (default false)"` part should be removed.
2. Cobra automatically adds a "(default <val>)" when it's something of interest (e.g not `""`).
3. Boolean flags should default to `false`. The absence of the flag as an arg should be the default behavior. The presence of a flag as an arg should change that behavior. When you default a boolean flag to `true`, that gets messed up; absence of the flag is now the same as providing it. In order to override the default, it would have to be provided as `--restricted=false` (you can't even do `--restricted false`).

TL;DR (too late!):
This should either default to `false`, or, if the same current behavior is desired, should be renamed to `--unrestricted` so that it can have a `false` default value. Either way the description should be cleaned up too.

See also: The `--iavl-disable-fastnode` debacle.

File: x/name/client/cli/gov.go [hunk 3]
```diff
@@ -106,7 +110,93 @@ $ %s tx gov submit-proposal param-change tx gov submit-proposal \
...
+func GetModifyNameProposalCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "modify-name-proposal [name] (--owner [address]) (--restrict) [flags]",
+		Short: "Submit a modify name creation governance proposal",
+		Long: strings.TrimSpace(
+			fmt.Sprintf(`Submit a modify name governance proposal along with an initial deposit.
+The proposal title and description must be provided through their respective flags.
+
+IMPORTANT: The creation of sub-names will be enabled by default unless the restrict flag is included.
+When included the owner must approve all child name creation unless an alterate owner is provided.
```
Comment (SpicyLemon) [inline x/name/client/cli/gov.go:131]: The `--restricted` flag is added to the command with a default of `true` (see other comment), so this blurb is inaccurate.

File: x/name/spec/04_events.md [hunk 4]
```diff
@@ -18,3 +19,13 @@ The name module emits the following events:
 | --------------------- | --------------------- | ------------------------- |
 | name_unbound          | name                  | {NameRecord|Name}         |
 | name_unbound          | address               | {NameRecord|Address}      |
+| name_unbound          | restricted            | {NameRecord|Restricted}   |
+
+
+### MsgBindNameRequest
```
Comment (SpicyLemon) [inline x/name/spec/04_events.md:25]:
```suggestion
### ModifyNameProposal
```

Also, does there need to be a section in here about events from a root name proposal?

**Review by SpicyLemon — COMMENTED** [review]
Couple small things, but mostly looks good.

File: x/name/client/cli/gov.go [hunk 5]
```diff
@@ -27,6 +27,10 @@ const (
 	FlagDeposit string = "deposit"
 )
 
+func GetProposalCmd() *cobra.Command {
+	return nil
+}
+
```
[the remark this answers is not in the document]
Comment (Taztingo) [reply x/name/client/cli/gov.go:33]: Whoops, forgot to remove this.

File: proto/provenance/name/v1/name.proto (hunk 1 above)
Comment (Taztingo) [reply proto/provenance/name/v1/name.proto:61]: Agreed

File: x/name/client/cli/gov.go (hunk 2 above)
Comment (Taztingo) [reply x/name/client/cli/gov.go:114]: I cleaned up the help text for this. Also, renamed the flag to `--unrestrict` in the BindName request, RootName proposal, and in this new proposal. I also made sure to mention this in the README as well.

File: x/name/spec/04_events.md (hunk 4 above)
Comment (Taztingo) [reply x/name/spec/04_events.md]: Yes, it does emit an event. So I added it.

File: x/name/keeper/keeper.go [hunk 6]
```diff
@@ -291,3 +294,34 @@ func isValidUUID(s string) bool {
 	}
 	return true
 }
+
+func (keeper Keeper) addRecord(ctx sdk.Context, name string, addr sdk.AccAddress, restrict, isModifiable bool) error {
+	key, err := types.GetNameKeyPrefix(name)
```
Comment (iramiller) [inline x/name/keeper/keeper.go:299]: I noticed that we still have amino in this module (albeit as a deprecated legacy compatibility path) ... this would be a good time to clean that up with a module upgrade method.  I am not sure how much longer amino will be present at all.

Comment (Taztingo) [reply x/name/keeper/keeper.go:299]: I created a separate issue for this since the removal of legacy amino code will affect the attribute module as well.

File: x/name/client/cli/gov.go [hunk 7]
```diff
@@ -106,8 +106,94 @@ $ %s tx gov submit-proposal param-change tx gov submit-proposal \
...
+IMPORTANT: The restricted creation of sub-names will be enabled by default unless the unrestricted flag is included.
+The owner must approve all child name creation unless an alterate owner is provided.
+
+Example:
+$ %s tx gov submit-legacy-proposal \
+	modify-name-proposal \
+	<name> \
+	--unrestrict  \ 
+	--owner <key_or_address> \
```
Comment (SpicyLemon) [inline x/name/client/cli/gov.go:134]: Nit: The keyring isn't referenced for this field, so this should just be `<address>`:
```suggestion
	--owner <address> \
```
It's also incorrectly stated as `<key_or_address>` in `GetRootNameProposalCmd`.

The `--from` field DOES use the keyring though, so it can stay as `<key_or_address>`.

File: proto/provenance/name/v1/name.proto [hunk 8]
```diff
@@ -40,11 +40,35 @@ message CreateRootNameProposal {
...
+  string name = 3;
+  // the address the name will resolve to
+  string owner = 4;
+  // a flag that indicates if an owner signature is required to add sub-names
+  bool restricted = 5;
+}
+
+// ModifyNameProposal details a proposal to update the address
+// and/or the restricted flag of a name.
+message ModifyNameProposal {
```
Comment (SpicyLemon) [inline proto/provenance/name/v1/name.proto:57]: I feel like we should do this as a `gov.v1`-style msg proposal instead of as `gov.v1beta1` content.

I know @llama-del-rey is working on updating our governance proposals to the `gov.v1` style, e.g. #1288. If we just do that right off the bat with this one, we don't end up with as many deprecated proto messages.

Does anyone else have thoughts on this?

Comment (Taztingo) [reply proto/provenance/name/v1/name.proto:57]: Sure, I can definitely convert it to that.

File: x/name/keeper/keeper.go [hunk 9]
```diff
@@ -291,3 +294,34 @@ func isValidUUID(s string) bool {
...
+		return err
+	}
+
+	store := ctx.KVStore(keeper.storeKey)
+	if store.Has(key) && !isModifiable {
+		return types.ErrNameAlreadyBound
+	}
+
+	record := types.NewNameRecord(name, addr, restrict)
+	if err = record.ValidateBasic(); err != nil {
```
Comment (nullpointer0x00) [inline x/name/keeper/keeper.go:310]: I'm wondering if we should start renaming these to just `Validate`.   Since, `ValidateBasic` is supposed to be for sdk.Msg types?

Comment (SpicyLemon) [reply x/name/keeper/keeper.go:315]: While it'd be nice to do that, it's probably better left for a PR doing only that for any applicable `ValidateBasic` functions.

File: x/name/keeper/msg_server.go [hunk 10]
```diff
@@ -148,3 +149,28 @@ func (s msgServer) DeleteName(goCtx context.Context, msg *types.MsgDeleteNameReq
 
 	return &types.MsgDeleteNameResponse{}, nil
 }
+
+// ModifyName updates an existing name
+func (s msgServer) ModifyName(goCtx context.Context, msg *types.MsgModifyNameRequest) (*types.MsgModifyNameResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	if msg.GetAuthority() != s.Keeper.GetAuthority() {
+		return nil, sdkerrors.ErrInvalidRequest.Wrap(govtypes.ErrInvalidSigner.Error())
```
Comment (SpicyLemon) [inline x/name/keeper/msg_server.go:158]:
```suggestion
		return nil, sdkerrors.ErrInvalidRequest.Wrap(govtypes.ErrInvalidSigner.Wrapf("expected %s got %s", s.Keeper.GetAuthority(), msg.GetAuthority()))
```

File: x/name/spec/03_messages.md [hunk 11]
```diff
@@ -52,6 +52,30 @@ This message is expected to fail if:
...
+  // The address signing the message
+  string authority = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // The record being updated
+  NameRecord record = 2 [(gogoproto.nullable) = false];
+}
+```
+
+This message is expected to fail if:
+- Any components of the request do not pass basic integrity and format checks
+- The record to remove does not exist
```
Comment (SpicyLemon) [inline x/name/spec/03_messages.md:73]:
```suggestion
- The record to update does not exist
```

File: x/name/spec/03_messages.md [hunk 12]
```diff
@@ -52,6 +52,30 @@ This message is expected to fail if:
...
+- The requestor does not match the owner listed on the record.
```
Comment (SpicyLemon) [inline x/name/spec/03_messages.md:74]:
```suggestion
```

This is inaccurate.

File: x/name/types/msgs.go [hunk 13]
```diff
@@ -101,3 +103,34 @@ func (msg MsgDeleteNameRequest) GetSigners() []sdk.AccAddress {
...
+}
+
+// ValidateBasic runs stateless validation checks on the message.
+func (msg MsgModifyNameRequest) ValidateBasic() error {
+	if strings.TrimSpace(msg.Record.Name) == "" {
+		return fmt.Errorf("name cannot be empty")
+	}
+	if strings.TrimSpace(msg.Record.Address) == "" {
+		return fmt.Errorf("address cannot be empty")
+	}
```
Comment (SpicyLemon) [inline x/name/types/msgs.go:122]: Should check if the `msg.Record.Address` is a valid Bech32 instead of just checking for empty/whitespace.
```suggestion
	if _, err := sdk. AccAddressFromBech32(msg.Record.Address); err != nil {
		return fmt.Errorf("invalid record address: %w", err)
	}
```
For the authority, though, it's okay to just do the empty string check since it's only ever used as a string.

File: x/name/types/proposal.go [hunk 14]
```diff
@@ -11,13 +11,15 @@ import (
...
 	ProposalTypeCreateRootName = "CreateRootName"
+	ProposalTypeModifyName     = "ModifyName"
 )
 
 // Assert CreateRootNameProposal implements govtypesv1beta1.Content at compile-time
 var _ govtypesv1beta1.Content = &CreateRootNameProposal{}
 
 func init() {
 	govtypesv1beta1.RegisterProposalType(ProposalTypeCreateRootName)
+	govtypesv1beta1.RegisterProposalType(ProposalTypeModifyName)
```
Comment (SpicyLemon) [inline x/name/types/proposal.go:22]: Don't need this since it doesn't use the v1beta1 stuff.

**Review by derekadams — APPROVED** [review]
LGTM

**Review by SpicyLemon — APPROVED** [review]

**Review by derekadams — APPROVED** [review]
